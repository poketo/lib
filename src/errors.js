// @flow

type ErrorCode =
  | 'INVALID_URL'
  | 'UNSUPPORTED_SITE'
  | 'UNSUPPORTED_SITE_REQUEST'
  | 'TIMEOUT'
  | 'HTTP_ERROR';

class PoketoError extends Error {
  code: ErrorCode;
  constructor(errorCode: ErrorCode, message: string) {
    super(message);
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.code = errorCode;
  }
}

class InvalidUrlError extends PoketoError {
  constructor(url: string) {
    super('INVALID_URL', `Unable to parse url '${url}'`);
  }
}

class HTTPError extends PoketoError {
  statusCode: number;
  url: string;

  constructor(statusCode: number, message: string, url: string) {
    super('HTTP_ERROR', message);
    this.statusCode = statusCode;
    this.url = url;
  }
}

class NotFoundError extends HTTPError {
  constructor(url: string) {
    super(404, 'Not Found', url);
  }
}

class TimeoutError extends PoketoError {
  constructor(message: string, url: string) {
    super('TIMEOUT', message);
  }
}

class UnsupportedSiteError extends PoketoError {
  constructor(site: string) {
    super('UNSUPPORTED_SITE', `Site at '${site}' is not supported`);
  }
}

class UnsupportedSiteRequestError extends PoketoError {
  constructor(siteName: string, operationName: string) {
    super(
      'UNSUPPORTED_SITE_REQUEST',
      `${siteName} does not support ${operationName}`,
    );
  }
}

export default {
  PoketoError,
  HTTPError,
  InvalidUrlError,
  NotFoundError,
  TimeoutError,
  UnsupportedSiteError,
  UnsupportedSiteRequestError,
};
