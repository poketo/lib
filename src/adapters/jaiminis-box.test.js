import site from './jaiminis-box';
import errors from '../errors';

describe('JaiminisBoxAdapter', () => {
  describe('supportsUrl', () => {
    it('returns true for urls like jaiminisbox.com', () => {
      expect(site.supportsUrl('http://jaiminisbox.com')).toBe(true);
      expect(site.supportsUrl('http://jaiminisbox.com/a/path')).toBe(true);
      expect(site.supportsUrl('https://www.jaiminisbox.com/another?path')).toBe(
        true,
      );
    });

    it('returns false for urls that are not jaiminisbox.com', () => {
      expect(site.supportsUrl('http://he.lveticascans.com')).toBe(false);
      expect(site.supportsUrl('http://mangaupdates.com')).toBe(false);
    });
  });

  describe('parseUrl', () => {
    it('returns the components of a url', () => {
      expect(
        site.parseUrl('https://jaiminisbox.com/reader/series/itoshi-no-muco/'),
      ).toEqual({ seriesSlug: 'itoshi-no-muco', chapterSlug: null });

      expect(
        site.parseUrl(
          'https://jaiminisbox.com/reader/read/itoshi-no-muco/en/1/2/page/1',
        ),
      ).toEqual({ seriesSlug: 'itoshi-no-muco', chapterSlug: 'en/1/2' });

      expect(
        site.parseUrl(
          'https://jaiminisbox.com/reader/read/mousou-telepathy/en/0/512/5/page/25',
        ),
      ).toEqual({ seriesSlug: 'mousou-telepathy', chapterSlug: 'en/0/512/5' });
    });

    it('throws on unparseable paths', () => {
      expect(() => {
        site.parseUrl(
          'https://jaiminisbox.com/reader/other/mousou-telepathy/en/2/11/page/1',
        );
      }).toThrow(errors.InvalidUrlError);

      expect(() => {
        site.parseUrl('https://jaiminisbox.com/reader/');
      }).toThrow(errors.InvalidUrlError);
    });
  });

  describe('getSeries', () => {
    it('returns a metadata object', async () => {
      const metadata = await site.getSeries('itoshi-no-muco');

      expect(metadata).toEqual({
        slug: 'itoshi-no-muco',
        url: 'https://jaiminisbox.com/reader/series/itoshi-no-muco',
        title: 'Itoshi no Muco',
        chapters: expect.arrayContaining([
          {
            chapterNumber: '4',
            volumeNumber: '0',
            createdAt: 1483516800,
            slug: 'en/0/4',
            title: 'Walk!',
            url: 'https://jaiminisbox.com/reader/read/itoshi-no-muco/en/0/4/',
          },
          {
            chapterNumber: '8',
            volumeNumber: '0',
            createdAt: 1487750400,
            slug: 'en/0/8',
            title: 'Friend!',
            url: 'https://jaiminisbox.com/reader/read/itoshi-no-muco/en/0/8/',
          },
        ]),
      });
    });
  });

  describe('getChapter', () => {
    it('returns a list of pages', async () => {
      const chapter = await site.getChapter('itoshi-no-muco', 'en/0/4');

      expect(chapter.pages).toHaveLength(8);
      expect(chapter.pages).toEqual(
        expect.arrayContaining([
          {
            id: expect.any(Number),
            width: expect.any(Number),
            height: expect.any(Number),
            url: expect.stringContaining('http'),
          },
        ]),
      );
    });
  });
});
