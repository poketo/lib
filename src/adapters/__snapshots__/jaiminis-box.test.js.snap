// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JaiminisBox getSeries returns series metadata 1`] = `
Object {
  "author": "MIZUSHINA Takayuki",
  "coverImageUrl": "https://i0.wp.com/jaiminisbox.com/reader/content/comics/itoshi-no-muco_586c3e495e85d/thumb_mucothumbnailresized.jpg",
  "description": "The happy daily life of Muco, a not-too-bright Akita dog, and his owner Komatsu-san the glass-blower. Based on a real dog of the same name. (Mangaupdates)
",
  "id": "jaiminis-box:itoshi-no-muco",
  "site": Object {
    "id": "jaiminis-box",
    "name": "Jaiminiâ€™s Box",
  },
  "slug": "itoshi-no-muco",
  "status": "UNKNOWN",
  "supportsReading": true,
  "title": "Itoshi no Muco",
  "updatedAt": Any<Number>,
  "url": "http://localhost:57154/reader/series/itoshi-no-muco",
}
`;

exports[`JaiminisBox getSeries returns series metadata 2`] = `
Array [
  Object {
    "chapterNumber": "8",
    "createdAt": 1487750400,
    "id": "jaiminis-box:itoshi-no-muco:en/0/8",
    "order": 4,
    "slug": "en/0/8",
    "title": "Friend!",
    "url": "https://jaiminisbox.com/reader/read/itoshi-no-muco/en/0/8/",
    "volumeNumber": "0",
  },
  Object {
    "chapterNumber": "4",
    "createdAt": 1483516800,
    "id": "jaiminis-box:itoshi-no-muco:en/0/4",
    "order": 0,
    "slug": "en/0/4",
    "title": "Walk!",
    "url": "https://jaiminisbox.com/reader/read/itoshi-no-muco/en/0/4/",
    "volumeNumber": "0",
  },
]
`;

exports[`JaiminisBox getSeries returns sub-chapter metadata 1`] = `
Array [
  Object {
    "chapterNumber": "101.1",
    "createdAt": 1532502000,
    "id": "jaiminis-box:kaguya-wants-to-be-confessed-to:en/0/101/1",
    "order": 13,
    "slug": "en/0/101/1",
    "title": "Vol 10 Extras",
    "url": "https://jaiminisbox.com/reader/read/kaguya-wants-to-be-confessed-to/en/0/101/1/",
    "volumeNumber": "0",
  },
  Object {
    "chapterNumber": "101",
    "createdAt": 1521270000,
    "id": "jaiminis-box:kaguya-wants-to-be-confessed-to:en/0/101",
    "order": 12,
    "slug": "en/0/101",
    "title": "The Student Council Wants to be Photographed",
    "url": "https://jaiminisbox.com/reader/read/kaguya-wants-to-be-confessed-to/en/0/101/",
    "volumeNumber": "0",
  },
]
`;
